src/flamegraph/color/palette_map.rs

```diff
@@ -5,67 +5,96 @@ use std::fs::OpenOptions;
+/// from this function.
+#[derive(Default)]
+pub struct PaletteMap<'a>(HashMap<Cow<'a, str>, (u8, u8, u8)>);
 
 impl<'a> PaletteMap<'a> {
-    pub fn load(file: &str) -> quick_xml::Result<Self> {
+    /// Returns the color value corresponding to the given function name.
+    pub fn get(&self, func: &str) -> Option<(u8, u8, u8)> {
+        self.0.get(func).cloned()
+    }
+
+    /// Inserts a function name/color pair in the map.
+    pub fn insert(&mut self, func: &'a str, color: (u8, u8, u8)) -> Option<(u8, u8, u8)> {
+        self.0.insert(Cow::from(func), color)
+    }
+
+    /// Provides an iterator over the elements of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&str, (u8, u8, u8))> {
+        self.0.iter().map(|(func, color)| (func.as_ref(), *color))
+    }
+
+    /// Builds a mapping based on the inputs given by the reader.
+    ///
+    /// The reader should provide name/color pairs as text input, each pair separated by a line
+    /// separator.
+    ///
+    /// Each line should follow the format: NAME->rgb(RED, GREEN, BLUE)
+    /// where NAME is the function name, and RED, GREEN, BLUE integer values between 0 and 255
+    /// included.
+    ///
+    /// This function will return an [`std::io::Error`] if the input is not correctly formatted.
+    pub fn from_stream(reader: &mut dyn io::Read) -> io::Result<Self> {
         let mut map = HashMap::default();
-        let path = Path::new(file);
+        let reader = BufReader::new(reader);
 
-        // If the file does not exist, it is probably the first call to flamegraph with a consistent
-        // palette: there is nothing to load.
-        if path.exists() {
-            let file = File::open(path).map_err(quick_xml::Error::Io)?;
-            let file = BufReader::new(file);
-
-            for line in file.lines() {
-                let line = line.map_err(quick_xml::Error::Io)?;
-
-                // A line is formatted like this: NAME -> rbg(RED, GREEN, BLUE)
-                let mut words = line.split("->");
-
-                let name = match words.next() {
-                    Some(name) => name,
-                    None => return Err(quick_xml::Error::UnexpectedToken(line)),
-                };
-
-                let color = match words.next() {
-                    Some(name) => name,
-                    None => return Err(quick_xml::Error::UnexpectedToken(line)),
-                };
-
-                if words.next().is_some() {
-                    return Err(quick_xml::Error::UnexpectedToken(line));
-                }
-
-                let rgb_color = parse_rgb_string(color)
-                    .ok_or_else(|| quick_xml::Error::UnexpectedToken(color.to_string()))?;
-                map.insert(Cow::from(name.to_string()), rgb_color);
-            }
+        for line in reader.lines() {
+            let line = line?;
+            let (name, color) = parse_line(&line)?;
```

Comment:
Would it be better to warn and skip rather than return if we encounter an ill-formatted line?

Reply:
I find it difficult to handle errors if we just print a warning. But the alternative, if we want to let the user choose what to do, is to accumulate the errors and return something like `(Self, Iterator<Item=io::Error>)` ? It may be a little too complex for what we want to do here.

Reply:
I was more thinking "be liberal in what you accept" here. If a file has one line with some garbage in it, it seems unfortunate to fail on the entire file?

Reply:
[Done](https://github.com/jonhoo/inferno/pull/77/commits/f99e779f3d876c95dac2e022186a1f4d89cc7cd3)